app/forms/questionnaire_form.py

```diff
@@ -25,29 +31,88 @@ def __init__(self, schema, block_json, formdata=None, **kwargs):
-        Validate this form as usual and check for any form-level date-range validation errors
+        Validate this form as usual and check for any form-level date-range/calculated validation errors
         :return:
         """
-        valid_form = True
         valid_fields = FlaskForm.validate(self)
+        valid_form = True
 
-        for question_id, period_from_id, period_to_id in self.date_ranges:
-            if period_from_id not in self.errors and period_to_id not in self.errors:
-                period_from = getattr(self, period_from_id)
-                period_to = getattr(self, period_to_id)
-                validator = DateRangeCheck()
+        for question in self.schema.get_questions_for_block(self.block_json):
+            if question['type'] == 'DateRange':
+                period_from_id = question['answers'][0]['id']
+                period_to_id = question['answers'][1]['id']
+                if not (
+                        self.answers_all_valid([period_from_id, period_to_id]) and
+                        self._validate_date_range_question(question['id'], period_from_id, period_to_id)):
+                    valid_form = False
+            elif question['type'] == 'Calculated':
+                target_answer = self.schema.get_answer(question['calculated']['answer_id'])
+                target_total = self.answer_store.filter(answer_ids=[target_answer['id']]).values()[0]
+                if not (
+                        self.answers_all_valid(question['calculated']['answers_to_calculate']) and
+                        self._validate_calculated_question(question, target_answer, target_total)):
+                    valid_form = False
+
+        return valid_form and valid_fields
+
+    def _validate_date_range_question(self, question_id, period_from_id, period_to_id):
+        period_from = getattr(self, period_from_id)
+        period_to = getattr(self, period_to_id)
+        validator = DateRangeCheck()
+
+        # Check every field on each form has populated data
+        populated = all(field.data for field in itertools.chain(period_from, period_to))
+
+        if populated:
+            try:
+                validator(self, period_from, period_to)
+            except validators.ValidationError as e:
+                self.question_errors[question_id] = str(e)
+                return False
+
+        return True
+
+    def _validate_calculated_question(self, question, target_answer, target_total):
+        calculated_question = question['calculated']
+        message = None
+        if 'validation' in question:
+            message = question['validation'].get('messages')
+
+        validator = SumCheck(messages=message, currency=target_answer.get('currency'))
+
+        calculation = self._get_calculation_type(calculated_question['calculated_type'])
+
+        formatted_values = self._get_formatted_calculation_values(calculated_question['answers_to_calculate'])
+
+        calculation_total = Decimal('0')
```

Comment:
Also since `calculation` and `formatted_values` is only used in this method this would be easier to understand if it was created in `_get_calculation_total`

Reply:
needs it as breakdown fields can be not mandatory hence user wont enter anything